cpp/src/arrow/compute/kernels/scalar_set_lookup_benchmark.cc

```diff
@@ -0,0 +1,143 @@
+// under the License.
+
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/kernels/test_util.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/util/benchmark_util.h"
+
+namespace arrow {
+namespace compute {
+
+constexpr auto kSeed = 0x94378165;
+
+static void SetLookupBenchmarkString(benchmark::State& state,
+                                     const std::string& func_name,
+                                     const int32_t valueSetLength) {
+  const int64_t array_length = 1 << 20;
+  const int64_t value_min_size = 0;
+  const int64_t value_max_size = 32;
+  const double null_probability = 0.01;
+  random::RandomArrayGenerator rng(kSeed);
+
+  auto values =
+      rng.String(array_length, value_min_size, value_max_size, null_probability);
+  auto value_set =
+      rng.String(valueSetLength, value_min_size, value_max_size, null_probability);
+  ABORT_NOT_OK(CallFunction(func_name, {values, value_set}));
+  for (auto _ : state) {
+    ABORT_NOT_OK(CallFunction(func_name, {values, value_set}));
+  }
+  state.SetItemsProcessed((state.iterations() * array_length) +
+                          (state.iterations() * valueSetLength));
+  state.SetBytesProcessed((state.iterations() * values->data()->buffers[2]->size()) +
+                          (state.iterations() * value_set->data()->buffers[2]->size()));
+}
+
+template <typename Type>
+static void SetLookupBenchmarkNumeric(benchmark::State& state,
+                                      const std::string& func_name,
+                                      const int32_t valueSetLength) {
+  const int64_t array_length = 1 << 20;
+  const int64_t value_min_size = 0;
+  const int64_t value_max_size = 32;
+  const double null_probability = 0.01;
+  random::RandomArrayGenerator rng(kSeed);
+
+  auto values =
+      rng.Numeric<Type>(array_length, value_min_size, value_max_size, null_probability);
+  auto value_set =
+      rng.Numeric<Type>(valueSetLength, value_min_size, value_max_size, null_probability);
+  ABORT_NOT_OK(CallFunction(func_name, {values, value_set}));
+  for (auto _ : state) {
+    ABORT_NOT_OK(CallFunction(func_name, {values, value_set}));
+  }
+  state.SetItemsProcessed((state.iterations() * array_length) +
+                          (state.iterations() * valueSetLength));
+  state.SetBytesProcessed((state.iterations() * values->data()->buffers[1]->size()) +
+                          (state.iterations() * value_set->data()->buffers[1]->size()));
+}
+
+static void IndexInStringSmall(benchmark::State& state) {
+  SetLookupBenchmarkString(state, "index_in_meta_binary", state.range(0));
+}
+
+static void IsInStringSmall(benchmark::State& state) {
+  SetLookupBenchmarkString(state, "is_in_meta_binary", state.range(0));
+}
+
+static void IndexInStringLarge(benchmark::State& state) {
+  SetLookupBenchmarkString(state, "index_in_meta_binary", 1 << 20);
```

Comment:
`1 << 20` sounds very large and irrealistic for a lookup set. How about `1 << 10` (which is already quite a lot of lookup strings).